src/Optimizer.ts

```diff
@@ -0,0 +1,147 @@
+import { Report, ReportElement } from './Models/Report';
+import { parse } from '@asyncapi/parser';
+import { RemoveComponents } from './Optimizers/RemoveComponents';
+import { MoveToComponents } from './Optimizers/MoveToComponents';
+import { ReuseComponents } from './Optimizers/ReuseComponents';
+import { Options } from './Models/Options';
+import YAML from 'yaml';
+import merge from 'merge-deep';
+import * as _ from 'lodash';
+import { ComponentProvider } from './ComponentProvider';
+/**
+ * this class is the starting point of the library.
+ * user will only interact with this class. here we generate different kind of reports using optimizers, apply changes and return the results to the user.
+ *
+ * @public
+ */
+export class Optimizer {
+  componentProvider!: ComponentProvider;
+  reuseComponentsReport: ReportElement[] = [];
+  removeComponentsReport: ReportElement[] = [];
+  moveToComponentsReport: ReportElement[] = [];
+  outputObject = {};
+  constructor(private YAMLorJSON: any) {
+
```

Comment:
Is some implementation missing here? Else why do you need this constructor?

Reply:
I will add support for JSON and JSON strings in the future. the implementation will go there. #15